pkg/minio/minio.go

```diff
@@ -322,6 +324,18 @@ func (m *Minio) createBucketIfNotExists(ctx context.Context, bucketName string)
 		return ErrMinioClientNotInitialized
 	}
 
+	if m.muMap == nil {
+		m.muMap = make(map[string]*sync.Mutex)
+	}
+
+	mu, ok := m.muMap[bucketName]
+	if !ok {
+		mu = &sync.Mutex{}
+		m.muMap[bucketName] = mu
+	}
+	mu.Lock()
+	defer mu.Unlock()
+
```

Comment:
We should have some map clean up code on the error path. 

you need to add a variable name to the return error like this. 
```go
func (m *Minio) createBucketIfNotExists(ctx context.Context, bucketName string) (err error) {
```
then you could add a defer statement
```suggestion
defer func() {
   if err != nil {
      delete(mu.Map, bucketname)
   }
} ()
```

This would protect us from accessing the map and expecting the bucket to exist when it doesn't due to an error.

Reply:
The deletion is not necessary as it is unlocked on return even if it is due to an error.

```go
defer mu.Unlock()
```

Currently there are max 2 buckets created in a complete knuu run. But clean up is usually a good thing.

Why do we need to have var name `err` in the signature?

Reply:
The issues wasn't that use of the lock, but protecting against the error case of trying to use a bucket in the map what failed to be created.

Reply:
In this instance, adding the var name for the return error ensures that we are checking that error during the defer func. Also where we would want to define the defer func, an error hasn't been defined yet, so we need to define one anyways.

Reply:
got it, in the proposed defer func, we wanted to cleanup the Map and then each call causes it to create a new entry for the same key which is not very optimized. So we need to decide on that. I prefer to keep it without defer for sake of optimization as that function is called for every file push.